Tools/speech_recognizer/GSSpeechRecognitionServer.m

```diff
@@ -0,0 +1,146 @@
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import "GSSpeechRecognitionServer.h"
+#import "GSSpeechRecognitionEngine.h"
+#import <Foundation/Foundation.h>
+
+static GSSpeechRecognitionServer *_sharedInstance;
+static int _clients = 0;
+
+@implementation GSSpeechRecognitionServer
+
+/**
+ * Monitor connection...
+ */
++ (void)connectionDied: (NSNotification*)aNotification
+{
+  NSArray *objs = [[aNotification object] localObjects];
+  NSEnumerator *en = [objs objectEnumerator];
+  id o = nil;
+
+  if(_clients > 0)
+    {
+      _clients--;
+    }
+  
+  if(_clients == 0)
+    {
+      NSLog(@"Client count is zero, exiting");
+      exit(0);
+    }
+  
+  NSLog(@"NSSpeechRecognizer server connection count = %d after disconnection", _clients);
+  while((o = [en nextObject]) != nil)
+    {
+      if ([o isKindOfClass: self])
+        {
+          RELEASE(o);
+        }
+    }
+}
+
++ (void)initialize
+{
+  _sharedInstance = [[self alloc] init];
+  _clients = 0;
+  [[NSNotificationCenter defaultCenter]
+		addObserver: self
+		   selector: @selector(connectionDied:)
+                       name: NSConnectionDidDieNotification
+                     object: nil];
+}
+
++ (void)start
+{
+  NSConnection *connection = [NSConnection defaultConnection];
+  [connection setRootObject: _sharedInstance];
+  RETAIN(connection);
+  if (NO == [connection registerName: @"GSSpeechRecognitionServer"])
+    {
+      NSLog(@"Could not register name, make sure another server is not running.");
+      return;
+    }
+  [[NSRunLoop currentRunLoop] run];
+}
+
++ (id)sharedServer
+{
+  NSLog(@"NSSpeechRecognizer server connection count = %d after connection", _clients);
+  return _sharedInstance;
+}
+
+- (void) addClient
+{
+  _clients++;
+}
+
+- (id)init
+{
+  if (nil == (self = [super init]))
+    {
+      return nil;
+    }
+  
+  _engine = [GSSpeechRecognitionEngine defaultSpeechRecognitionEngine];
+  if (nil == _engine)
+    {
+      [self release];
+      return nil;
+    }
+  else
+    {
+      NSLog(@"Got engine starting... %@", _engine);
+      [_engine start];
+    }
+
+  _blocking = [[NSMutableArray alloc] initWithCapacity: 10];  // 10 seems reasonable...
+
+  return self;
+}
+
+- (void) dealloc
+{
+  [_engine stop];
+  RELEASE(_engine);
+  RELEASE(_blocking);
+  [super dealloc];
+}
+
+- (void) addToBlockingRecognizers: (NSString *)s
+{
+  [_blocking addObject: s];
+}
+
+- (void) removeFromBlockingRecognizers: (NSString *)s
+{
+  [_blocking removeObject: s];
+}
+
+- (BOOL) isBlocking: (NSString *)s
+{
+  return [_blocking containsObject: s];
```

Comment:
As in Highlander there can only be one blocking all others. So the logic here should be that of a stack. The object at the top blocks all other recognizers. You should ask whether a recognizer is blocked, that would be the case if the stack is not empty and the recogniser is not at the top.

Reply:
I was always partial to Ramirez.... he had style... plus he was played by Sean Connery. ;) Anyway, I have changed this.